### PR TITLE
[#8011] Instructor: edit students: improve efficiency of 'delete all students'

### LINKED ISSUE #8011: Instructor: edit students: improve efficiency of 'delete all students'
v5.110

I noticed that an instructor is trying to 'delete all students' of a course with 500+ students and constantly running into a 'task cancelled' error.

```
teammates.common.util.Logger severe: Unexpected exception caught by ControllerServlet :  (Logger.java:52)
java.util.concurrent.CancellationException: Task was cancelled.
	at com.google.common.util.concurrent.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1170)
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:492)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:402)
	at com.google.common.util.concurrent.AbstractFuture$TrustedFuture.get(AbstractFuture.java:89)
	at com.google.appengine.tools.development.TimedFuture.get(TimedFuture.java:42)
	at com.google.common.util.concurrent.ForwardingFuture.get(ForwardingFuture.java:62)
	at com.google.appengine.tools.appstats.RecordingFuture.get(RecordingFuture.java:108)
	at com.google.appengine.tools.appstats.RecordingFuture.get(RecordingFuture.java:20)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:89)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$7.get(AsyncDatastoreServiceImpl.java:406)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$7.get(AsyncDatastoreServiceImpl.java:402)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:89)
	at com.googlecode.objectify.impl.ResultAdapter.now(ResultAdapter.java:34)
	at com.googlecode.objectify.util.ResultWrapper.translate(ResultWrapper.java:22)
	at com.googlecode.objectify.util.ResultWrapper.translate(ResultWrapper.java:10)
	at com.googlecode.objectify.util.ResultTranslator.nowUncached(ResultTranslator.java:21)
	at com.googlecode.objectify.util.ResultCache.now(ResultCache.java:30)
	at teammates.storage.api.EntitiesDb.deleteEntity(EntitiesDb.java:216)
	at teammates.logic.core.FeedbackResponsesLogic.deleteFeedbackResponseAndCascade(FeedbackResponsesLogic.java:682)
	at teammates.logic.core.FeedbackResponsesLogic.deleteFeedbackResponsesForStudentAndCascade(FeedbackResponsesLogic.java:746)
	at teammates.logic.core.StudentsLogic.deleteStudentCascade(StudentsLogic.java:476)
	at teammates.logic.core.StudentsLogic.deleteStudentCascade(StudentsLogic.java:467)
	at teammates.logic.core.StudentsLogic.deleteAllStudentsInCourse(StudentsLogic.java:462)
	at teammates.logic.api.Logic.deleteAllStudentsInCourse(Logic.java:1033)
	at teammates.ui.controller.InstructorCourseStudentDeleteAllAction.execute(InstructorCourseStudentDeleteAllAction.java:21)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:474)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:75)
	at teammates.ui.controller.ControllerServlet.doGet(ControllerServlet.java:50)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:143)
	at teammates.ui.controller.AppstatsFilter.doFilter(AppstatsFilter.java:41)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.googlecode.objectify.ObjectifyFilter.doFilter(ObjectifyFilter.java:48)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:48)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:257)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:76)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:146)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:657)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:619)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:589)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:453)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:460)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:293)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:319)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:311)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:457)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:263)
	at java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by josiahkhoo**
Fixes #8011

**Outline of Solution**

This pull request attempts to improve the efficiency of 'delete all students' by doing two things:
1. Parallelizing the retrieval of students, their respective feedback responses, and the comments for the responses.
2. Deleting students, feedback response and comments in batch.

Initially, I attempted to retrieve students, feedback responses and comments in batch using the `in` condition like so:

```
// Batch query using `objectify`
load()
	.filter("receiver in", emailList) 
	.list();
```

However, `objectify 6.0.7` does not allow for `in` conditions to be used in queries for non-key fields. In previous versions, `in` condition was supposedly a wrapper for performing multiple queries one after another. Thus, I opted to retrieve the student, feedback responses and comments in parallel.

### CODE REVIEW

**Comment by rrtheonlyone** [comment]
Good start here @josiahkhoo! Nice work.

Can you profile your change and show me the expected improvement in performance? You have to create a large class and compare how long the API takes to return in a staging environment (This is not easy though, @moziliar / @daongochieu2810 will be able to help you with this). 

At first glance, I think this method is still better: https://github.com/TEAMMATES/teammates/issues/8011#issuecomment-703785852 and will ensure there is no timeouts. This is mainly because step (2) is many API calls done for each feedback session. Your solution might still timeout for large courses.

**Comment by josiahkhoo** [comment]
> long waiting time on its own does not make bad UX. Long waiting time without any progress indicator is. Without any progress indicator, never mind 60 seconds: even 10-15 seconds is already (borderline) unacceptable UX.

@wkurniawan07 I see, what do you think about showing the user a loading wheel along with a small dialog that says "Currently deleting all students, please do not close this page (this can take sometime)" when they perform the 'delete all students' action? That gives the user some form of indication that this process can take some time + doesn't require having to keep track of deletion progress in the backend.

**Comment by damithc** [comment]
> 1. Deletion happens in the background, modal that alerts the user that this process might take a while pops out - click on `OK` to hide the modal, or leave it there till the deletion completes. **This is the part which was added.**
> 
> Screenshot 2021-10-15 at 18-22-12 TEAMMATES - Online Peer Feedback Evaluation System for Student Team Projects

Thanks for taking this on @josiahkhoo 

1. hmm... leaving the page doesn't actually affect the delete operation, right?
2. I'm not sure this additional modal helps that much, as opposed to a modal that shows a progress bar (or a spinner) and a message `deleting in progress, please wait ...` and automatically closes when the operation is done.

**Comment by josiahkhoo** [comment]
> hmm... leaving the page doesn't actually affect the delete operation, right?

Yeap leaving this page doesn't affect delete.

> I'm not sure this additional modal helps that much, as opposed to a modal that shows a progress bar (or a spinner) and a message deleting in progress, please wait ... and automatically closes when the operation is done.

Right, I can tweak the modal to do this! I am using the common modal component, so will extend that component to support what you said. (Since the current options for the common modal are quite limited - must have a primary button and the closed icon)

Thanks for the input @damithc :)

**Comment by damithc** [comment]
> Right, I can tweak the modal to do this! I am using the common modal component, so will extend that component to support what you said. (Since the current options for the common modal are quite limited - must have a primary button and the closed icon)

@josiahkhoo We already have a modal with a progress bar. I think that's in the page used to enroll students IIRC.

**Comment by damithc** [comment]
@josiahkhoo I'm fine with this approach. Thoughts from others welcome too.

File: src/main/java/teammates/logic/api/Logic.java
Comment (daongochieu2810) [inline src/main/java/teammates/logic/api/Logic.java:758]: I feel like fetching the whole object just to delete it is unnecessary. We can just get the id

Comment (moziliar) [reply src/main/java/teammates/logic/api/Logic.java:758]: I agree. Is it possible to do key only query to get the id and proceed to delete by id?

File: src/web/app/components/simple-modal/simple-modal.component.spec.ts
Comment (moziliar) [inline src/web/app/components/simple-modal/simple-modal.component.spec.ts:17]: Any reason for this indentation?

File: src/web/app/components/simple-modal/simple-modal.module.ts
Comment (moziliar) [inline src/web/app/components/simple-modal/simple-modal.module.ts:4]: Was the progress-bar.module somehow repositioned?

File: src/test/java/teammates/storage/api/FeedbackResponsesDbTest.java
Comment (moziliar) [inline src/test/java/teammates/storage/api/FeedbackResponsesDbTest.java:378]: Any reason for this too?

**Review by moziliar — CHANGES_REQUESTED** [review]
It seems that you are parallelising object fetch. This is a good direction to hide the latency.

Do note that you are effectively getting the object key before deletion. It is worth researching if you can perform key only query to save the whole object fetch and the key create overhead.

File: src/test/java/teammates/storage/api/FeedbackResponsesDbTest.java
Comment (josiahkhoo) [reply src/test/java/teammates/storage/api/FeedbackResponsesDbTest.java]: This is me over-linting, I reverted this change.

File: src/web/app/components/simple-modal/simple-modal.component.spec.ts
Comment (josiahkhoo) [reply src/web/app/components/simple-modal/simple-modal.component.spec.ts]: This is me over-linting, I reverted this change.

File: src/web/app/components/simple-modal/simple-modal.module.ts
Comment (josiahkhoo) [reply src/web/app/components/simple-modal/simple-modal.module.ts:4]: Yeap, sorted the ordering after adding `AjaxLoadingModule` because TSLint complains when the import is not sorted.

File: src/main/java/teammates/logic/api/Logic.java
Comment (josiahkhoo) [reply src/main/java/teammates/logic/api/Logic.java:770]: Updated the functions to perform a query to only retrieve the `id` for Student, FeedbackResponse and FeeedbackResponseComment (`email` and `courseId` for Student) using `project` method from the objectify [package](https://www.javadoc.io/static/com.googlecode.objectify/objectify/6.0/com/googlecode/objectify/cmd/Query.html#project-java.lang.String...-).

**Review by daongochieu2810 — APPROVED** [review]

File: src/web/app/components/simple-modal/simple-modal.module.ts
Comment (moziliar) [reply src/web/app/components/simple-modal/simple-modal.module.ts:4]: Ah nvm. I see that the changed line actually got a redundant path. This is alright.

**Comment by moziliar** [comment]
@josiahkhoo As per discussion, perhaps we should think about how the deletion workflow can bypass the 1min timeout limit for the user?
 
At where it stands right now, we can only allow x2 of the student deletion. Perhaps you can update the stats here after the key-only query change?

It might be a plus to show the staging server dashboard on the data read data here for evaluation, as those operations cost money whereas memory read doesn't.

**Comment by josiahkhoo** [comment]
Just ran the LNP script on the updated function which uses projection queries. This is the result when trying to delete 1000 students:
```
[=============================teammates.lnp.cases.InstructorStudentDeleteLNPTest=============================]

#Req: 1,  Throughput: 0.02/s,  Avg resp time: 54.82s,  90th percentile: 54.82s,  Err: 0 (0.00%)
 L&P results for teammates.lnp.cases.InstructorStudentDeleteLNPTest:
 >
 > #Req: 1,  Throughput: 0.02/s,  Avg resp time: 54.82s,  90th percentile: 54.82s,  Err: 0 (0.00%)
lnp-tests > lnp-tests > teammates.lnp.cases.InstructorStudentDeleteLNPTest > runLnpTest PASSED

```
For 1000 students, the response time has reduced from ~90s to ~50s.

This is the memory usage in comparison to the previous implementation:

New Project

**Comment by moziliar** [comment]
@damithc @rrtheonlyone Any opinion on the stats shown? Looks a little expensive to me.

I am personally ok with not having the extra index (if @josiahkhoo has added it that caused the memory increase) on feedback response and comments. The entity read and write will be an one-off cost.

**Comment by damithc** [comment]
Thanks for the work so far @josiahkhoo and @moziliar This is not a straight-forward issue to tackle.

I was under the impression that this was just a matter of 'keeping the user informed'? Given how rare deleting a big class is, I prefer a simpler solution if there is one.

> @josiahkhoo As per discussion, perhaps we should think about how the deletion workflow can bypass the 1min timeout limit for the user?

I thought the limit is 10 minutes, after the V8 migration? Or was I mistaken?

Let's wait to hear from @wkurniawan07 and @rrtheonlyone too, as they have an in-depth view of the code complexity aspect.

**Comment by jianhandev** [comment]
Trying to look into this issue/PR, meanwhile @rrtheonlyone @wkurniawan07 would appreciate your inputs on this PR.

**Comment by hhdqirui** [comment]
@josiahkhoo kind reminder on this PR.

**Comment by josiahkhoo** [comment]
In this case would it be better to tweak the 1min limit for this particular API? Since extremely large classes might still exceed the 1min limit in spite of optimisation as.

**Comment by jianhandev** [comment]
@josiahkhoo thanks for the work so far, as you have correctly identified, extremely large classes might still exceed the 1-min limit. As this is a rather complicated performance issue, it deserves more discussion within the core team. I'll put this PR on hold first until we come up with a more scalable solution.

**Comment by wkurniawan07** [comment]
Closing as the issue has been resolved.

**Pull request closed without merging**
